lib/DxilValidation/DxilValidation.cpp

```diff
@@ -970,6 +970,226 @@ static void ValidateImmOperandForMathDxilOp(CallInst *CI, DXIL::OpCode Opcode,
+    return;
+  }
+
+  ConstantInt *IsOutputUnsignedConst =
+      dyn_cast<llvm::ConstantInt>(IsOutputUnsigned);
+  if (!IsOutputUnsignedConst) {
+    ValCtx.EmitInstrError(CI,
+                          ValidationRule::InstrMatVecOpIsUnsignedFlagsAreConst);
+    return;
+  }
+
+  ConstantInt *II = dyn_cast<ConstantInt>(InputInterpretation);
+  ConstantInt *MI = dyn_cast<ConstantInt>(MatrixInterpretation);
+  if (!II || !MI) {
+    ValCtx.EmitInstrError(
+        CI, ValidationRule::InstrLinalgInterpretationParamAreConst);
+    return;
+  }
+
+  // Check if InputInterpretation and MatrixInterpretation are valid
+  auto IIValue = II->getLimitedValue();
+  if (!CheckFromRegisterInterpretations(IIValue)) {
+    ValCtx.EmitInstrError(
+        CI, ValidationRule::InstrLinalgInvalidRegisterInterpValue);
+    return;
+  }
+
+  auto MIValue = MI->getLimitedValue();
+  if (!CheckInMemoryInterpretations(MIValue)) {
+    ValCtx.EmitInstrError(CI,
+                          ValidationRule::InstrLinalgInvalidMemoryInterpValue);
+    return;
+  }
+
+  ConstantInt *MatrixTransposeConst = dyn_cast<ConstantInt>(MatrixTranspose);
+  ConstantInt *MatrixLayoutConst = dyn_cast<ConstantInt>(MatrixLayout);
+  if (!llvm::isa<llvm::Constant>(MatrixM) ||
+      !llvm::isa<llvm::Constant>(MatrixK) || !MatrixLayoutConst ||
+      !MatrixTransposeConst) {
+    ValCtx.EmitInstrError(CI,
+                          ValidationRule::InstrLinalgMatrixShapeParamsAreConst);
+    return;
+  }
+
+  auto MLValue = MatrixLayoutConst->getLimitedValue();
+  if (!CheckMatrixLayout(MLValue)) {
+    ValCtx.EmitInstrError(CI,
+                          ValidationRule::InstrLinalgInvalidMatrixLayoutValue);
+    return;
+  }
+
+  if (!CheckTransposeForMatrixLayout(
+          static_cast<DXIL::DXILMatrixLayout>(MLValue),
+          MatrixTransposeConst->getLimitedValue())) {
+    ValCtx.EmitInstrError(
+        CI, ValidationRule::InstrLinalgMatrixLayoutNotTransposable);
+    return;
+  }
+
+  llvm::Value *InputVector =
+      CI->getOperand(DXIL::OperandIndex::kMatVecMulInputVectorIdx);
+  if (!CheckUnsignedFlag(InputVector->getType(),
+                         IsInputUnsignedConst->getLimitedValue())) {
+    ValCtx.EmitInstrError(CI, ValidationRule::InstrLinalgNotAnUnsignedType);
+    return;
+  }
+
+  if (!CheckUnsignedFlag(CI->getType(),
+                         IsOutputUnsignedConst->getLimitedValue())) {
+    ValCtx.EmitInstrError(CI, ValidationRule::InstrLinalgNotAnUnsignedType);
+    return;
+  }
+
+  switch (OpCode) {
+  case DXIL::OpCode::MatVecMulAdd: {
+    llvm::Value *BiasInterpretation =
+        CI->getOperand(DXIL::OperandIndex::kMatVecMulAddBiasInterpretation);
+    if (!llvm::isa<llvm::Constant>(BiasInterpretation)) {
+      ValCtx.EmitInstrError(
+          CI, ValidationRule::InstrLinalgInterpretationParamAreConst);
+      return;
+    }
+    ConstantInt *BI = cast<ConstantInt>(BiasInterpretation);
+    auto BIValue = BI->getLimitedValue();
```

Comment:
nit: https://llvm.org/docs/CodingStandards.html#use-auto-type-deduction-to-make-code-more-readable
```suggestion
    uint64_t BIValue = BI->getLimitedValue();
```